metrics-scraper/monitoring.go

```diff
@@ -69,16 +70,49 @@ func Start(configPath string) {
 	wg.Wait()
 }
 
+// tryConnect, try to connect to the node forever until it works
+// it'll stop when it acquire a connection or it's asked to stop
+func tryConnect(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	config NodeConfig,
+) (c *rpc.Client, err error) {
+	c, err = rpc.DialContext(ctx, config.RPCURL)
+	if err == nil {
+		// no errors, let's start polling
+		return
+	}
+
+	ticker := time.NewTicker(5 * time.Second) // arbitrarily retry every 5 seconds
+	for {
+		select {
+		case <-ctx.Done():
+			wg.Done()
+			log.Printf("job terminated for %v", config.Moniker)
+			return
+		case <-ticker.C:
+			c, err = rpc.DialContext(ctx, config.RPCURL)
+			if err != nil {
+				log.Printf("error: couldn't connect to %v at %v: %v", config.Moniker, config.RPCURL, err)
+				continue
+			}
+
+			return
+		}
+	}
+}
+
 func run(
 	ctx context.Context,
 	wg *sync.WaitGroup,
 	pollRate time.Duration,
 	networkID string,
 	config NodeConfig,
 ) {
-	c, err := rpc.DialContext(ctx, config.RPCURL)
+	c, err := tryConnect(ctx, wg, config)
```

Comment:
I think there is no need to pass `wg` down to `tryConnect`. It's enough to do `defer wg.Done()` at the beginning of `run` to handle all cases. If we do that. we can also remove the `wg.Done()` call from the `<-ctx.Done()` case within this method.

Reply:
addressed in: d1755782d3115b584b67e6fb763dc3d62386ef81